pkg/test/local/environment.go

```diff
@@ -15,38 +15,62 @@
 package local
 
 import (
+	"fmt"
 	"testing"
 
+	"istio.io/istio/pkg/test/dependency"
 	"istio.io/istio/pkg/test/environment"
 	"istio.io/istio/pkg/test/internal"
 )
 
 // Environment a local environment for testing. It implements environment.Interface, and also
 // hosts publicly accessible methods that are specific to local environment.
 type Environment struct {
-	ctx internal.TestContext
+	ctx *internal.TestContext
 }
 
 var _ environment.Interface = &Environment{}
-var _ Internal = &Environment{}
+var _ internal.Environment = &Environment{}
 
 // NewEnvironment returns a new instance of local environment.
-func NewEnvironment(ctx internal.TestContext) (*Environment, error) {
-	return &Environment{
-		ctx: ctx,
-	}, nil
+func NewEnvironment() (*Environment, error) {
+	return &Environment{}, nil
+}
+
+// Initialize the environment. This is called once during the lifetime of the suite.
+func (e *Environment) Initialize(ctx *internal.TestContext) error {
+	e.ctx = ctx
+	return nil
+}
+
+// InitializeDependency is called when a new dependency is encountered during test run.
+func (e *Environment) InitializeDependency(ctx *internal.TestContext, d dependency.Instance) (interface{}, error) {
+	switch d {
+	// TODO
+	default:
+		return nil, fmt.Errorf("unrecognized dependency: %v", d)
+	}
 }
 
 // Configure applies the given configuration to the mesh.
 func (e *Environment) Configure(t testing.TB, config string) {
-	// TODO
-	panic("Not yet implemented")
+	for _, d := range e.ctx.Tracker().All() {
+		if configurable, ok := d.(internal.Configurable); ok {
+			err := configurable.ApplyConfig(config)
```

Comment:
AFAICT nothing currently implements `Configurable`.  Which dependencies do you envision implementing it and how will this work? I'm guessing that `config` will be the configuration for the entire mesh (k8s yaml?)? If so, why would we provide the entire system configuration to each dependency? Will each `Configurable` entity be able to parse out its own piece of configuration from the mesh config?

Reply:
I don't see any other way. The alternative is to apply configuration to each component directly, which is a bit awkward, and possibly makes the authored test more susceptible to our changes.   We can build enough smarts in the framework (without being too smart) that can easily deal with this.

Reply:
And to answer your first question, it will be done as "service" of the tracker: The local environment will traverse and apply config to all configurables. The configurables can pick the parts that are interesting to them and apply. In the cluster case, the environment would directly apply using Kubectl.

Reply:
Looks like the config will be parsed every time it passed to the ApplyConfig method. Will it make sense to parse it once in a struct entity, and transfer to configurables?

Reply:
I think it is pretty cheap (cpu/memory-wise) to reparse config. It will be more expensive (development-wise) to maintain a set of structured entities for this purpose.